[#40397] OpenAPI Specs not OpenAPI 3.0 compatible

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/40397
- fixed crucial bugs and most warnings

Hopefully the number of changed files doesn't scare you ^^. Basically there are only a handful of unique changes and fixes, and then numerous insertions of the same schema into dozens of response specifications.

After those changes, the `openapitools/openapi-generator-cli` runs fine, only a certain number of warnings remain, but no fatal errors.

Things, I noticed, which could be worth improving:

1. Almost all examples of media-type-objects (the ones within response specifications) are wrong, as they are nested within a `response.value` path. Here is how media-type-examples should look like: https://spec.openapis.org/oas/v3.1.0#media-type-examples
2. there is a script https://github.com/opf/openproject/blob/dev/script/api/validate_spec, which obviously didn't find the three real validation errors. It might come in handy to look, why our validation didn't find those. The fatal errors before those changes of mine were:

```
Errors: 
        -attribute components.schemas.Link.examples is unexpected
        -attribute paths.'/api/v3/views/{identifier}'. Declared path parameter identifier needs to be defined as a path parameter in path or operation level
        -attribute paths.'/api/v3/views/{id}'(get).responses.200.description is missing

```

Kharonus committed:
[#40397] OpenAPI Specs not OpenAPI 3.0 compatible

- https://community.openproject.org/work_packages/40397
- fixed crucial bugs and most warnings
- fixed spelling and naming

machisuji commented:
@Kharonus Thanks for all the effort! As far as I can tell the script does show some errors. But it shows 7 for me:

```
[1] docs/api/apiv3/openapi-spec.yml:365:3 at #/paths/~1api~1v3~1views~1{identifier}

The path already exists which differs only by path parameter name(s): `/api/v3/views/{id}` and `/api/v3/views/{identifier}`.

363 | "/api/v3/views/{id}":
364 |   "$ref": "./paths/view.yml"
365 | "/api/v3/views/{identifier}":
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
366 |   "$ref": "./paths/views_{identifier}.yml"
367 | "/api/v3/wiki_pages/{id}":

Error was generated by the no-identical-paths rule.


[2] docs/api/apiv3/paths/groups.yml:166:25 at #/post/requestBody/content/application~1json/schema/properties/_links/properties/members/type

`type` can be one of the following only: "object", "array", "string", "number", "integer", "boolean", "null".

Did you mean: array ?

164 | properties:
165 |   members:
166 |     type: arrray
    |           ^^^^^^
167 |     items:
168 |       type: object

Error was generated by the spec rule.


[3] docs/api/apiv3/paths/group.yml:248:25 at #/patch/requestBody/content/application~1json/schema/properties/_links/properties/members/type

`type` can be one of the following only: "object", "array", "string", "number", "integer", "boolean", "null".

Did you mean: array ?

246 | properties:
247 |   members:
248 |     type: arrray
    |           ^^^^^^
249 |     items:
250 |       type: object

Error was generated by the spec rule.


[4] docs/api/apiv3/paths/view.yml:13:5 at #/get/responses/200

The field `description` must be present on this level.

11 |       type: integer
12 | responses:
13 |   '200':
   |   ^^^^^
14 |     content:
15 |       application/hal+json:

Error was generated by the spec rule.


[5] docs/api/apiv3/paths/views_{identifier}.yml:29:29 at #/post/requestBody/content/application~1json/schema/properties/_links/properties/query/properties/href/type

`type` can be one of the following only: "object", "array", "string", "number", "integer", "boolean", "null".

27 |                 properties:
28 |                   href:
29 |                     type: href
   |                           ^^^^
30 | responses:
31 |   '201':

Error was generated by the spec rule.


[6] docs/api/apiv3/paths/views_{identifier}.yml:4:1 at #/post/parameters

The operation does not define the path parameter `{identifier}` expected by path `/api/v3/views/{identifier}`.

2 | ---
3 |
4 | post:
  | ^^^^
5 |   requestBody:
6 |     content:

Error was generated by the path-parameters-defined rule.


[7] docs/api/apiv3/components/schemas/user_preference_model.yml:12:11 at #/properties/notifications/type

`type` can be one of the following only: "object", "array", "string", "number", "integer", "boolean", "null".

10 |   description: Hide mail address from other users
11 | notifications:
12 |   type: NotificationSetting
   |         ^^^^^^^^^^^^^^^^^^^
13 |   description: The settings for the notifications to be received by the user
14 | timeZone:

Error was generated by the spec rule.
```

The used tool is supposed to be able to handle specs split up into files. But when I fully assemble the spec before testing it it results in even more errors. Can you confirm this?

I've created a PR (https://github.com/opf/openproject/pull/10032) which updates the validation script so that it 

1) fully assembles the spec before linting
2) returns the correct exit code in the end

It could be that PRs introduced errors because the github action didn't fail due to the script not returning the correct exit code at the end.

Kharonus commented:
Thanks @machisuji for pointing to that script, I fixed the npx call in the validation script, so it can now run on a system that does not have the `@redocly/openapi-cli` npm package installed.

And I fixed the 5 errors, that were still there after my changes.

Kharonus committed:
[#40397] added model for notification settings

Kharonus committed:
[#40397] fixed linter errors

- added model for notification settings
- reverted --yes option in nx call

Kharonus committed:
[#40397] fixed 2 more errors in code generator
